src/Microsoft.AspNet.Server.Kestrel/Http/FrameResponseStream.cs

```diff
@@ -112,24 +119,36 @@ public void StopAcceptingWrites()
         public void Abort()
         {
             // We don't want to throw an ODE until the app func actually completes.
-            // If the request is aborted, we throw an IOException instead.
             if (_state != FrameStreamState.Closed)
             {
                 _state = FrameStreamState.Aborted;
             }
         }
 
-        private void ValidateState()
+        private Task ValidateState(CancellationToken cancellationToken)
         {
             switch (_state)
             {
                 case FrameStreamState.Open:
-                    return;
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        return TaskUtilities.GetCancelledTask(cancellationToken);
+                    }
+                    break;
                 case FrameStreamState.Closed:
                     throw new ObjectDisposedException(nameof(FrameResponseStream));
                 case FrameStreamState.Aborted:
-                    throw new IOException("The request has been aborted.");
+                    if (cancellationToken.CanBeCanceled)
+                    {
+                        // Aborted state only throws on write if cancellationToken requests it
+                        return TaskUtilities.GetCancelledTask(
+                            cancellationToken.IsCancellationRequested ? 
+                            cancellationToken : 
+                            new CancellationToken(true));
```

Comment:
Pre-newer CancellationToken changes, needs to only throw if given `CancellationToken` is cancelled otherwise it should return `null`.